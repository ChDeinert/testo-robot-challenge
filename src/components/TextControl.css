.textcontrol {
  margin: -1rem 0 -.5rem;
  padding: .5rem;
  border: 2px solid var(--color-blue);
  border-radius: .25rem;
}

.textcontrol__form {
  display: flex;
  gap: 1rem;
  justify-content: center;
  align-items: center;
  margin: 1rem;
}

.textcontrol__form > label {
  font-weight: 600;
}

.textcontrol__form__input {
  padding: .25rem;
  background-color: white;
  border: 2px solid var(--color-blue);
  border-radius: .25rem;
}

.textcontrol__form__input[type="submit"] {
  cursor: pointer;
  padding-left: .5rem;
  padding-right: .5rem;
  font-weight: 600;
}

.textcontrol__form__input[type="submit"]:active,
.textcontrol__form__input[type="submit"]:hover {
  background-color: var(--color-blue-light);
}

.textcontrol__form__input:focus {
  border: 2px solid var(--color-blue);
  outline: var(--color-orange) solid 2px;
  outline-offset: 2px;
}

.textcontrol__message {
  min-height: calc(1.8rem + 4px);
  line-height: 1rem;
  padding: .25rem .5rem;
  border: 2px solid var(--color-red);
  border-radius: .25rem;
  text-align: center;
}

.textcontrol__message:empty {
  border-color: transparent;
}