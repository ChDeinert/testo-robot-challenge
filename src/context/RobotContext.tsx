import React, { Component, createContext, useContext } from 'react';
import Robot, { defaultBoard, possibleDirections, DirectionType, BoardType, PositionType } from '../lib/robot';

type RobotContextType = { 
  settings: {
    board: BoardType,
    possibleDirections: DirectionType[],
  },
  state: {
    initialized: boolean,
    isPlaced: boolean,
    position: { xPosition: number, yPosition: number, direction: DirectionType },
    error?: Error,
  },
  controls?: {
    place: (xCoordinate: number, yCoordinate: number, direction: DirectionType) => void,
    move: () => void,
    turnLeft: () => void,
    turnRight: () => void,
    report: () => undefined|PositionType,
  },
};

const defaultRobotContext : RobotContextType = {
  settings: {
    board: defaultBoard,
    possibleDirections: possibleDirections,
  },
  state: {
    initialized: false,
    isPlaced: false,
    position: { xPosition: 0, yPosition: 0, direction: possibleDirections[0] },
  },
};
const RobotContext = createContext<RobotContextType>(defaultRobotContext);
const useRobotContext = () => useContext(RobotContext);

/**
 * React component rendering the RobotContext provider,
 * it's setting, state and control-methods
 */
class RobotProvider extends Component {
  private robot?: Robot;
  state = {
    initialized: false,
    position: { xPosition: 0, yPosition: 0, direction: possibleDirections[0] },
    isPlaced: false,
    error: undefined
  };

  componentDidMount() {
    // Initialize the Robot only after the component is mounted
    // so that ssr/prerender aren't effected by any instance of Robot
    this.robot = new Robot(defaultBoard, possibleDirections);
    this.setState({ initialized: true });
  }

  updateCurrentPosition = () => {
    this.setState({ position: this.robot?.report(), error: undefined });
  }

  place = (xCoordinate: number, yCoordinate: number, facing: DirectionType) => {
    try {
      this.robot?.place(xCoordinate, yCoordinate, facing);
      this.updateCurrentPosition();
      this.setState({ isPlaced: true });
    } catch (error) {
      this.setState({ error });
    }
  }

  move = () => {
    try {
      this.robot?.move();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  turnLeft = () => {
    try {
      this.robot?.left();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  turnRight = () => {
    try {
      this.robot?.right();
      this.updateCurrentPosition();
    } catch (error) {
      this.setState({ error });
    }
  };

  report = () => {
    try {
      const currentPositionReport = this.robot?.report();
      alert(`The Robot's current position: ${currentPositionReport}`);
      return currentPositionReport;
    } catch (error) {
      this.setState({ error });
    }
  }

  render() { 
    const { children } = this.props;
    const { initialized, position, isPlaced, error } = this.state;

    return (
      <RobotContext.Provider value={{
        settings: { board: defaultBoard, possibleDirections: possibleDirections },
        state: {
          initialized,
          isPlaced,
          position,
          error,
        },
        controls: {
          place: this.place,
          move: this.move,
          turnLeft: this.turnLeft,
          turnRight: this.turnRight,
          report: this.report,
        }
      }}>
        {children}
      </RobotContext.Provider>
    );
  }
}

export default RobotContext;
export { RobotProvider, useRobotContext };