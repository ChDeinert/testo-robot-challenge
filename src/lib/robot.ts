type BoardType = { xWidth: number, yWidth: number };
type DirectionType = 'NORTH' | 'EAST' | 'SOUTH' | 'WEST';
type PositionType = { xPosition?: number, yPosition?: number, direction?: DirectionType, toString?: () => string };

/**
 * Represents the default Board that the Robot can be placed on
 */
const defaultBoard : BoardType = {
  xWidth: 5,
  yWidth: 5,
};

/**
 * Array containing the possible directions the robot can be facing
 */
const possibleDirections = [
  'NORTH' as DirectionType, 'EAST' as DirectionType, 'SOUTH' as DirectionType, 'WEST' as DirectionType
];

/**
 * Represents the Robot on the board,
 * it's containing the logic and provides the methods to control it
 */
class Robot {
  private board: BoardType;
  private possibleDirections : DirectionType[];
  private xPosition?: number;
  private yPosition?: number;
  private direction?: DirectionType;

  constructor(board: BoardType, possibleDirections: DirectionType[]) {
    this.board = board;
    this.possibleDirections = possibleDirections;
  }

  /**
   * Validates if a given x Coordinate is a valid one
   */
  private isValidXCoordinate = (xCoordinate: number) : boolean => {
    if (typeof xCoordinate !== 'number' || isNaN(xCoordinate) || xCoordinate < 0 || xCoordinate >= this.board.xWidth) {
      return false;
    }

    return true;
  };

  /**
   * Validates if a given y Coordinate is a valid one
   */
  private isValidYCoordinate = (yCoordinate: number) : boolean => {
    if (typeof yCoordinate !== 'number' || isNaN(yCoordinate) || yCoordinate < 0 || yCoordinate >= this.board.yWidth) {
      return false;
    }

    return true;
  };

  /**
   * Validates if a given direction string is a valid one
   */
  private isValidDirection = (direction: DirectionType) : boolean => {
    if (typeof direction !== 'string' || !this.possibleDirections.includes(direction)) {
      return false;
    }

    return true;
  };

  /**
   * Validates if the robot is positioned on the board
   */
  private isRobotPlaced = (xCoordinate?: number, yCoordinate?: number, direction?: DirectionType) : boolean => {
    if (typeof xCoordinate !== 'number' || !this.isValidXCoordinate(xCoordinate) || 
        typeof yCoordinate !== 'number' || !this.isValidYCoordinate(yCoordinate) || 
        typeof direction === 'string' && !this.isValidDirection(direction)) {
      return false;
    }

    return true;
  };

  /**
   * Returns the next facing direction of the Robot depending on the turning direction and it's current facing direction
   * 
   * @throws Error - Errors if the Coordinates or Direction are invalid
   */
  private nextDirection = (turnDirection: 'left'|'right') : undefined|DirectionType => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        return turnDirection === 'left' ? 'WEST' : 'EAST';
      case 'EAST':
        return turnDirection === 'left' ? 'NORTH' : 'SOUTH';
      case 'SOUTH':
        return turnDirection === 'left' ? 'EAST' : 'WEST';
      case 'WEST':
        return turnDirection === 'left' ? 'SOUTH' : 'NORTH';
    }
  };

  /**
   * Placing the Robot on the Board on the given Coordinates and facing the given direction
   * 
   * @throws Error - Errors if the Coordinates or Direction are invalid
   */
  place = (x: number, y: number, facing: DirectionType) : void => {
    if (!this.isValidXCoordinate(x)) {
      throw new Error('X Coordinate is invalid');
    }
    if (!this.isValidYCoordinate(y)) { 
      throw new Error('Y Coordinate is invalid');
    }
    if (!this.isValidDirection(facing)) {
      throw new Error('The given direction, the Robot is facing, is invalid');
    }

    this.xPosition = x;
    this.yPosition = y;
    this.direction = facing;
  };

  /**
   * Moves the robot on the Board
   * 
   * @throws Error - Errors if the Robot is not placed yet or the new Coordinates are out of bounds
   */
  move = () : void => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        if (typeof this.yPosition !== 'number' || !this.isValidYCoordinate(this.yPosition + 1)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        ++this.yPosition;
        break;
      case 'EAST':
        if (typeof this.xPosition !== 'number' || !this.isValidXCoordinate(this.xPosition + 1)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        ++this.xPosition;
        break;
      case 'SOUTH':
        if (typeof this.yPosition !== 'number' || !this.isValidYCoordinate(this.yPosition - 1)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        --this.yPosition;
        break;
      case 'WEST':
        if (typeof this.xPosition !== 'number' || !this.isValidXCoordinate(this.xPosition - 1)) {
          throw new Error('The robot can\'t move in that direction, or it falls down');
        }
        --this.xPosition;
        break;
    }
  };

  /**
   * Turnes the Robot to the left (counter clockwise)
   */
  left = () : void => {
    this.direction = this.nextDirection('left');
  };

  /**
   * Turnes the Robot to the right (clockwise)
   */
  right = () : void => {
    this.direction = this.nextDirection('right');
  };

  /**
   * Reports the Robot's current position
   * 
   * @throws Error - Errors if the Robot is not placed yet
   */
  report = () : PositionType => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    return {
      xPosition: this.xPosition,
      yPosition: this.yPosition,
      direction: this.direction,
      toString: () => `${this.xPosition}, ${this.yPosition}, ${this.direction}`,
    };
  };
}

export default Robot;
export {
  defaultBoard,
  possibleDirections,
  BoardType,
  PositionType,
  DirectionType,
};