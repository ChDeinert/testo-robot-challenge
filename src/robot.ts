type Board = { xWidth: number, yWidth: number };
type Direction = 'NORTH' | 'EAST' | 'SOUTH' | 'WEST';
type Position = { xPosition: undefined|number, yPosition: undefined|number, direction: undefined|Direction, toString?: () => string };

const defaultBoard : Board = {
  xWidth: 5,
  yWidth: 5,
};

const possibleDirections = [
  'NORTH', 'EAST', 'SOUTH', 'WEST'
];

const isRobotPlaced = (board : Board, possibleDirections: Direction[]) => (xCoordinate : undefined|number, yCoordinate : undefined|number, direction: undefined|Direction) : boolean => {
  if (!isValidXCoordinate(xCoordinate, board) || !isValidYCoordinate(yCoordinate, board) || !isValidDirection(direction, possibleDirections)) {
    return false;
  }

  return true;
};

const isValidXCoordinate = (xCoordinate : number, board: Board) : boolean => {
  if (typeof xCoordinate !== 'number' || xCoordinate < 0 || xCoordinate >= board.xWidth) {
    return false;
  }

  return true;
};

const isValidYCoordinate = (yCoordinate : number, board: Board) : boolean => {
  if (typeof yCoordinate !== 'number' || yCoordinate < 0 || yCoordinate >= board.yWidth) {
    return false;
  }

  return true;
};

const isValidDirection = (direction : Direction, possibleDirections: Direction[]) : boolean => {
  if (typeof direction !== 'string' || !possibleDirections.includes(direction)) {
    return false;
  }

  return true;
};

class Robot {
  private xPosition: undefined | number;
  private yPosition: undefined | number;
  private direction: undefined | Direction;
  private board: Board;
  private possibleDirections : Direction[];
  private isRobotPlaced;

  constructor(board: Board, possibleDirections : Direction[]) {
    this.board = board;
    this.possibleDirections = possibleDirections;
    this.isRobotPlaced = isRobotPlaced(board, possibleDirections);
  }

  place = (x: number, y: number, facing: Direction) : void => {
    if (!isValidXCoordinate(x, this.board)) {
      throw new Error('X Coordinate is invalid');
    }
    if (!isValidYCoordinate(y, this.board)) { 
      throw new Error('Y Coordinate is invalid');
    }
    if (!isValidDirection(facing, this.possibleDirections)) {
      throw new Error('The given direction, the Robot is facing, is invalid');
    }

    this.xPosition = x;
    this.yPosition = y;
    this.direction = facing;
  };

  move = () : void => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }
  };

  left = () : void => {
    this.direction = this.nextDirection('left');
  };

  right = () : void => {
    this.direction = this.nextDirection('right');
  };

  report = () : Position => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    return {
      xPosition: this.xPosition,
      yPosition: this.yPosition,
      direction: this.direction,
      toString: () => `${this.xPosition}, ${this.yPosition}, ${this.direction}`,
    };
  };

  private nextDirection = (turnDirection : 'left'|'right') : Direction => {
    if (!this.isRobotPlaced(this.xPosition, this.yPosition, this.direction)) {
      throw new Error('The robot is not placed yet');
    }

    switch (this.direction) {
      case 'NORTH':
        return turnDirection === 'left' ? 'WEST' : 'EAST';
      case 'EAST':
        return turnDirection === 'left' ? 'NORTH' : 'SOUTH';
      case 'SOUTH':
        return turnDirection === 'left' ? 'EAST' : 'WEST';
      case 'WEST':
        return turnDirection === 'left' ? 'SOUTH' : 'NORTH';
    }
  };
};

export default Robot;
export { defaultBoard, possibleDirections, isRobotPlaced, isValidXCoordinate, isValidYCoordinate, isValidDirection };