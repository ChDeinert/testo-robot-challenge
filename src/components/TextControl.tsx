import React from 'react';
import { useForm, SubmitHandler } from "react-hook-form";
import { useRobotContext } from '../context/RobotContext';
import { DirectionType } from '../lib/robot';
import './TextControl.css';

const TextControl : React.FC = () => {
  const { state: { initialized }, controls } = useRobotContext();
  const { register, handleSubmit, formState: { errors } } = useForm();

  const handleCommandSubmit: SubmitHandler<{ command: string }> = ({ command }) => {
    const currentCommand = command.trim().match(/^(\w+)\(/);

    switch (currentCommand && currentCommand[1]) {
      case 'place':
        const placeParams = command.trim().match(/\((\d+),\s?(\d+),\s?'?(NORTH|EAST|SOUTH|WEST)'?\)$/);
        const xCoordinate = (placeParams && parseInt(placeParams[1])) ?? -1;
        const yCoordinate = (placeParams && parseInt(placeParams[2])) ?? -1;
        const facing = (placeParams && placeParams[3]) ?? '';
        controls?.place(xCoordinate, yCoordinate, facing as DirectionType);
        break;
      case 'move':
        controls?.move();
        break;
      case 'left':
        controls?.turnLeft();
        break;
      case 'right':
        controls?.turnRight();
        break;
      case 'report':
        controls?.report();
        break;
    }
  };

  const validateCommand = (commandValue: string) => {
    if (!commandValue.trim().match(/^place\(\d+,\s?\d+,\s?'?(?:NORTH|EAST|SOUTH|WEST)'?\)$|^move\(\)$|^left\(\)$|^right\(\)$|^report\(\)$/)) {
      return `The command "${commandValue}" is invalid!`;
    }
    
    return true;
  };

  return !initialized ? (<p>loading...</p>) : (
    <details className="textcontrol">
      <summary>Manually type Commands for the robot</summary>
      <form onSubmit={handleSubmit(handleCommandSubmit)} className="textcontrol__form" data-testid="robotcommandform">
        <label htmlFor="textcontrol-command">
          Command: 
        </label>
        <input 
          type="text" 
          id="textcontrol-command"
          {...register('command', { required: 'A command is required', validate: validateCommand })} 
          className="textcontrol__form__input" 
          data-testid="robotcommand-input" 
        />
        <input type="submit" value="Send command" className="textcontrol__form__input" data-testid="robotcommand-submit" />
      </form>
      <div className="textcontrol__message">
        {errors?.command && (<span>{errors?.command.message}</span>)}
      </div>
    </details>
  );
};

export default TextControl;