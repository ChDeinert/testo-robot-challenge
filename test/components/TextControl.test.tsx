import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import TextControl from "../../src/components/TextControl";
import RobotContext from "../../src/context/RobotContext";
import { defaultBoard, possibleDirections } from "../../src/lib/robot";

describe('<TextControl />', () => {
  it('renders a loading indication if Robot is not initialized in context', async () => {
    render(<TextControl />);
    const placeForm = screen.getByText('loading...');
    expect(placeForm).toMatchInlineSnapshot(`
      <p>
        loading...
      </p>
    `);
  });

  it('renders the command form', () => {
    const testContext = {
      settings: {
        board: defaultBoard,
        possibleDirections: possibleDirections,
      },
      state: {
        initialized: true,
        isPlaced: false,
        position: {
          xPosition: 0,
          yPosition: 0,
          direction: possibleDirections[0],
        },
      },
    };
    render(
      <RobotContext.Provider value={testContext}>
        <TextControl />
      </RobotContext.Provider>
    );
    const placeForm = screen.getByTestId('robotcommandform');
    expect(placeForm).toMatchSnapshot();
  });
});
