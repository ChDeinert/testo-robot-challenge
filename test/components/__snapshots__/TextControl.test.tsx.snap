// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextControl /> renders the command form 1`] = `
<form
  class="textcontrol__form"
  data-testid="robotcommandform"
>
  <label
    for="textcontrol-command"
  >
    Command:
  </label>
  <input
    class="textcontrol__form__input"
    data-testid="robotcommand-input"
    id="textcontrol-command"
    name="command"
    type="text"
  />
  <input
    class="textcontrol__form__input"
    data-testid="robotcommand-submit"
    type="submit"
    value="Send command"
  />
</form>
`;
